app/javascript/controllers/map_controller.js

```diff
@@ -0,0 +1,96 @@
+import { Controller } from "stimulus"
+import consumer from "../channels/consumer"
+
+export default class extends Controller {
+  static targets = ["image"]
+
+  connect() {
+    this.mapId = this.element.dataset.mapId
+    this.channel = consumer.subscriptions.create({
+      channel: "MapChannel",
+      id: this.mapId
+    }, {
+      received: this.cableReceived.bind(this)
+    })
+  }
+
+  moveMap(event) {
+    const speedFactor = 2
+    var mousedown = {
+      x: event.pageX,
+      y: event.pageY
+    }
+    var backgroundPosition = {
+      x: Number.parseInt(event.target.style.backgroundPosition.split(" ")[0]),
+      y: Number.parseInt(event.target.style.backgroundPosition.split(" ")[1])
+    }
+    var handleMove = (event) => {
+      var mousepos = {
+        x: event.pageX,
+        y: event.pageY
+      }
+      if (mousedown != mousepos) {
+        var moveby = {
+          x: (mousepos.x - mousedown.x) * speedFactor,
+          y: (mousepos.y - mousedown.y) * speedFactor
+        }
+        this.channel.perform(
+          "move_map",
+          {
+            map_id: this.mapId,
+            x: backgroundPosition.x + moveby.x,
+            y: backgroundPosition.y + moveby.y
+          }
+        )
+      }
+    }
+    document.addEventListener("mouseup", function () {
+      document.removeEventListener("mousemove", handleMove)
+    })
+    document.addEventListener("mousemove", handleMove)
+  }
+
+  setMapPosition(x, y) {
+    this.imageTarget.style.backgroundPosition = `${x}px ${y}px`
+  }
+
+  zoom(factor) {
+    var backgroundSize = {
+      width: Number.parseInt(this.imageTarget.style.backgroundSize.split(" ")[0]),
+      height: Number.parseInt(this.imageTarget.style.backgroundSize.split(" ")[1])
+    }
+    this.channel.perform(
+      "set_size",
+      {
+        map_id: this.mapId,
+        width: Math.round(backgroundSize.width * factor),
+        height: Math.round(backgroundSize.height * factor),
+      }
+    )
+  }
+
+  zoomIn(event) {
+    event.preventDefault()
+    this.zoom(1.25)
+  }
+
+  zoomOut(event) {
+    event.preventDefault()
+    this.zoom(0.75)
```

Comment:
Just thought - I think these zoom values mean you can't get back to 1.

```
1 * 0.75 * 1.25  = 0.9375
1 * 0.8  * 1.25  = 1
1 * 0.75 * (4/3) = 1
```

Reply:
Good catch, you're right. I'm going to change the way zooming works to solve this and other issues.